Services/Simulations.cs

```diff
@@ -107,38 +114,53 @@ public async Task<IList<Models.Simulation>> GetListAsync()
         public async Task<Models.Simulation> GetAsync(string id)
         {
             var item = await this.storage.GetAsync(STORAGE_COLLECTION, id);
-            var simulation = JsonConvert.DeserializeObject<Models.Simulation>(item.Data);
-            simulation.ETag = item.ETag;
-            return simulation;
+
+            if (item != null)
+            {
+                var simulation = JsonConvert.DeserializeObject<Models.Simulation>(item.Data);
+                simulation.ETag = item.ETag;
+                simulation.Id = item.Key;
+                return simulation;
+            }
+
+            return null;
         }
 
         /// <summary>
         /// Create a simulation.
         /// </summary>
         public async Task<Models.Simulation> InsertAsync(Models.Simulation simulation, string template = "")
         {
-            // TODO: complete validation
-            if (!string.IsNullOrEmpty(template) && template.ToLowerInvariant() != "default")
+            var usingDefaultTemplate = !string.IsNullOrEmpty(template) && template.ToLowerInvariant() == DEFAULT_TEMPLATE_NAME;
+
+            if (!string.IsNullOrEmpty(template) && template.ToLowerInvariant() != DEFAULT_TEMPLATE_NAME)
             {
                 this.log.Warn("Unknown template name", () => new { template });
                 throw new InvalidInputException("Unknown template name. Try 'default'.");
             }
 
+            if (!usingDefaultTemplate && string.IsNullOrEmpty(simulation.Name))
+            {
+                this.log.Warn("Missing simulation name", () => new { });
+                throw new InvalidInputException("Simulation name is required.");
+            }
+
             var simulations = await this.GetListAsync();
-            if (simulations.Count > 0)
+            var activeSimulation = simulations.Where(a => a.ShouldBeRunning());
+            if (activeSimulation.Any())
             {
-                this.log.Warn("There is already a simulation");
+                this.log.Warn("There is already a running simulation");
```

Comment:
as per discussion, we're removing this, correct? so multiple simulations can be created, then the Agent temporarily will execute only one?

Reply:
removed